### PR TITLE
Add CEP and CES resources

### PR DESCRIPTION
**Pull request by pippolo84**
This is a required step to migrate the k8s CEP watchers to Resource[T].

First, the PR adds a separate constructor for the CiliumSlimEndpoint resource in the clustermesh-apiserver. The apiserver is currently using the same constructor used by the agent and this limits the ability to add indexers that depend on agent-specific part.
Then, the PR adds the CES resource constructor and the same indexers already used by the informers in the CEP and CES k8s watchers.

Refer to [this comment](https://github.com/cilium/cilium/issues/28159#issuecomment-1816498303) and to the individual commit messages for more information.

Related: https://github.com/cilium/cilium/issues/28159

### CODE REVIEW

File: clustermesh-apiserver/clustermesh/k8s/resource_ctors.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+)
+
+func CiliumSlimEndpointResource(lc hive.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*types.CiliumEndpoint], error) {
```
Comment (tommyp1ckles) [inline clustermesh-apiserver/clustermesh/k8s/resource_ctors.go:20]: might be worth adding a comment explaining why we have a clustermesh specific constructor here.

File: pkg/k8s/resource_ctors.go [hunk 2]
```diff
@@ -324,6 +324,40 @@ func ciliumEndpointLocalPodIndexFunc(obj interface{}) ([]string, error) {
...
+	}
+	return resource.New[*cilium_api_v2alpha1.CiliumEndpointSlice](lc, lw,
+		resource.WithMetric("CiliumEndpointSlice"),
+		resource.WithIndexers(indexers),
+	), nil
+}
+
+// ciliumEndpointSliceLocalPodIndexFunc is an IndexFunc that indexes CiliumEndpointSlices
+// by their corresponding Pod, which are running locally on this Node.
+func ciliumEndpointSliceLocalPodIndexFunc(obj interface{}) ([]string, error) {
```
Comment (tommyp1ckles) [inline pkg/k8s/resource_ctors.go]: nit: `obj any`

**Review by tommyp1ckles — APPROVED** [review]

**Review by YutaroHayakawa — APPROVED** [review]

File: pkg/k8s/resource_ctors.go (hunk 2 above)
Comment (pippolo84) [reply pkg/k8s/resource_ctors.go]: Fixed, thanks!

File: clustermesh-apiserver/clustermesh/k8s/resource_ctors.go (hunk 1 above)
Comment (pippolo84) [reply clustermesh-apiserver/clustermesh/k8s/resource_ctors.go:20]: Added, thanks!

**Comment by pippolo84** [comment]
@tommyp1ckles @YutaroHayakawa I've made a small change to the `CiliumSlimEndpointResource` and `CiliumEndpointSliceResource` constructors. Since both uses an IndexFunc that relies on the local node reference to be initialized, I've added `*LocalNodeStore` in their function signature. This is needed to make the dependency from the resources to the local node store explicit to the "hive and cells" framework.
Without that, it would have been possible to build a Hive with one or both the above resources provided and end up panicking while accessing the uninitialized local node.

**Pull request merged**